### PR TITLE
Sync issues, PRs, labels, milestones, comments from GitHub mirrors

### LINKED ISSUE #18369: Sync/Mirror Issues and Pull-Requests on migrated repository
### Feature Description

**Feature:** Want to sync/mirror Issue/PR on demand or timely interval.
**Current situation:** When migration (remote to gitea) is done and items like Issues/PR are checked, they are pulled perfectly initially. However, it's one-off and there is no way to fetch new issues/PR later on.
**Workaround:** One has to once again initiate a new migration whenever they need to pull new issues/PR which is not idle and efficient. Another solution would be to use Mirror feature, but that only allows branches, tags, and commits.
**Why:** It will help me to be in sync with my remote repository all the time and if in future I want to go Gitea exclusive then it can be done easily.

Version: Gitea Version: 1.15.10
PS: You guys are amazing btw and thank you for all the work and and providing this amazing project for free. :heart: 

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by harryzcy**
This PR adds support for one-way syncing of topics, milestones, labels, issues, pull requests, comments, reviews, releases, and reactions from pull mirror (GitHub) to Gitea.
Closes #18369.

`Downloader` interface is extended with `GetNewIssues`, `GetNewPullRequests`, since new issues and pull requests should be appended/updated to the existing repository.

`Uploader` interface is extended with `UpdateTopics`, `UpdateMilestones`, `UpdateLabels`, `PatchReleases`, `PatchComments`, `PatchIssues`, and `PatchPullRequests`. All `Update*` should replace existing data and `Patch*` should append/update but should not delete existing data.

The above methods will be implement here for GitHub, ~or possibly also GitLab,~ in this pull request.

In order to make the updates work, we would need to uniquely identify a Topic, Milestone, Label, Release, Comments, Issue and PullRequest. The following list shows how they are identified:

- [x] Topic - Name (it's just a string, unrelated to any other things)
- [x] Milestone - OriginalID*
- [x] Label - OriginalID*
- [x] Release - OriginalID*
- [x] Comments - OriginalID*
- [x] Issue - IssueID and Index
- [x] PullRequest - IssueID and Index
- [x] Reviews - OriginalID*
- [x] Reactions - IssueID, CommentID and Type

OriginalID*: additional column added to database, used to track ID from the source

Screenshot of toggles to set which migration Items should be synced (the only change is now when mirror option is checked, the migration items are still available for supported platforms):

image

### CODE REVIEW

**Comment by lunny** [comment]
You need to implement `UpdateIssues`, `UpdateReactions`, `UpdateComments` and etc. functions. You need to find the unique external id to update them.

File: services/migrations/gitea_uploader.go [hunk 1]
```diff
@@ -801,6 +802,39 @@ func (g *GiteaLocalUploader) CreateReviews(reviews ...*base.Review) error {
 	return issues_model.InsertReviews(cms)
 }
 
+// UpdateTopics updates topics
+func (g *GiteaLocalUploader) UpdateTopics(topics ...string) error {
+	return g.CreateTopics(topics...)
```
Comment (lunny) [inline services/migrations/gitea_uploader.go]: But maybe you need to delete the unused topics.

Comment (harryzcy) [reply services/migrations/gitea_uploader.go]: ~~Sure, I'll work on that~~
`repo_model.SaveTopics` actually already adds new topics and deletes unused ones

File: modules/migration/null_uploader.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migration
+
+// NullUploader implements a blank uploader
+type NullUploader struct{}
+
+var _ Downloader = &NullDownloader{}
```
Comment (delvh) [inline modules/migration/null_uploader.go]:
```suggestion
var _ Uploader = &NullUploader{}
```

File: models/migrate.go [hunk 3]
```diff
@@ -92,6 +98,97 @@ func insertIssue(ctx context.Context, issue *issues_model.Issue) error {
...
+	issueLabels := resolveIssueLabels(issue.ID, issue.Labels)
+	if len(issueLabels) > 0 {
+		// delete old labels
+		if _, err := sess.Where("issue_id = ?", issue.ID).Delete(); err != nil {
+			return err
+		}
+		// insert new labels
+		if _, err := sess.Insert(issueLabels); err != nil {
+			return err
+		}
```
Comment (harryzcy) [inline models/migrate.go]: Here we can do:
- delete all issueLabels and then recreate them (current approach)
- use another query beforehand to get existing issueLabels, then delete ones that should be deleted and insert ones that should be inserted

I'm not sure which approach works better. Since number of labels associated with an issue is likely small, I'm guessing delete-all-and-recreate approach will have a smaller overhead than an additional query. But still, I would like to know what everyone thinks.

Comment (harryzcy) [reply models/migrate.go]: ~~I think this is wrong and will change them later~~

**Comment by tecosaur** [comment]
@harryzcy this looks really promising! What's the current state of this?

**Comment by harryzcy** [comment]
It's on hold since I discovered PR #20122, which is basically doing the same thing. I don't know the status of that PR. So if there's an interest from the community in this PR, I can resume the work. @tecosaur

**Comment by nikoPLP** [comment]
@earl-warren proposes a new approach https://github.com/go-gitea/gitea/pull/20122#issuecomment-1236438810 and the PR #20122 you mentioned above is now closed. How do you see it @harryzcy in the near future?

**Comment by harryzcy** [comment]
> @earl-warren proposes a new approach [#20122 (comment)](https://github.com/go-gitea/gitea/pull/20122#issuecomment-1236438810) and the PR #20122 you mentioned above is now closed. How do you see it @harryzcy in the near future?

I will pick it up sometime.

**Comment by lukas-h** [comment]
any timeline on that feature?x

**Comment by harryzcy** [comment]
This PR is mostly complete and ready for reviews. There maybe some minor issues I'm looking into (updated release doesn't have the right database record, etc.), but that won't change the codebase much.

File: models/issues/label.go [hunk 4]
```diff
@@ -206,6 +208,10 @@ func NewLabels(labels ...*Label) error {
 
 // UpdateLabel updates label information.
 func UpdateLabel(l *Label) error {
+	return updateLabel(db.DefaultContext, l)
```
Comment (lunny) [inline models/issues/label.go]: Expose context is better.

Comment (harryzcy) [reply models/issues/label.go]: The original code uses `db.DefaultContext`. The whole purpose of lower case `updateLabel ` is to expose the context, so that upper case `UpdateLabel` can use the default context.

There are lots of code using the default context, so I'm inclined to keep it. Otherwise, it should be the job of a separate  PR, and many more functions should be updated all together.

**Comment by mpeter50** [comment]
I'm a bit afraid to ask because I don't completely understand the changes, but what will happen to references of issues and comments?

As I understand, if between 2 synchronizations a new issue is made both in the local copy of the repository and also at the remote one (e.g. Github), then after the next syncronization the new issue of the remote side will have a different ID locally than there, because they are appended to the list of existing ones.
This would mean that if that issue was referenced (by its ID, because thats how it's done) in an other issue, in a PR, in a commit, or somwehere else, then that reference would now point to a different issue than originally intended.
This actually also applies to PRs and comments, as they all have serially incremented IDs (and for comments, counting is done globally on the instance level afaik, not on the repository level).

Basically, what happens if there are objects at the remote repository that have the same IDs as some local obejcts?
If they get a new ID (because they are appended to existing ones or for an other reason), are references somehow corrected, or will they just become silently broken?

**Comment by harryzcy** [comment]
@mpeter50 No, two way synchronization is not supported in this PR. It is expected that users doesn't change anything on Gitea that's synced from GitHub, otherwise things will just get broken or get replaced.

Probably some resources should be made read-only, i.e. disable the buttons on the web and return 400 for APIs.

Or even better, we should start storing unique IDs from the remote (GitHub) in this PR so that it will be much easier to enable two-way syncs in the future.

Since this PR will get merged only after 1.20 release, there's more time to discuss about these to avoid premature implementations.

**Comment by harryzcy** [comment]
I have picked up this PR recently. Some of the pending tasks:

- when comment is updated on GitHub, the updated time in Gitea is not updated (probably have the same issue with other things)
- release can't be correctly updated

Some non blocking issues:
- comment history from GitHub not imported

File: services/migrations/migrate.go [hunk 5]
```diff
@@ -312,9 +314,9 @@ func migrateRepository(doer *user_model.User, downloader base.Downloader, upload
 		}
 
 		// Once all releases (if any) are inserted, sync any remaining non-release tags
-		if err = uploader.SyncTags(); err != nil {
-			return err
-		}
+		// if err = uploader.SyncTags(); err != nil {
+		// 	return err
+		// }
```
Comment (harryzcy) [inline services/migrations/migrate.go]: It seems this code is causing issues with release syncs. We got all the releases already and then tries to sync "any remaining non-release tags", which is not the case.

It's actually syncing all the tags including release tags by calling `SyncReleasesWithTags`, and even worse, it's deleting ALL the releases and recreating them in a broken way:

If repo is a mirror, `pullMirrorReleaseSync` in `modules/repository/repo.go` is called. Then all releases are deleted, and releases are recreated from tags without any `Author`, `NumCommits`, etc. This is the wrong behavior because we do need the release information.

```go
for _, tag := range tags {
	release := repo_model.Release{
		RepoID:       repo.ID,
		TagName:      tag.Name,
		LowerTagName: strings.ToLower(tag.Name),
		Sha1:         tag.Object.String(),
		// NOTE: ignored, since NumCommits are unused
		// for pull-mirrors (only relevant when
		// displaying releases, IsTag: false)
		NumCommits:  -1,
		CreatedUnix: timeutil.TimeStamp(tag.Tagger.When.Unix()),
		IsTag:       true,
	}
	if err := db.Insert(ctx, release); err != nil {
		return fmt.Errorf("unable insert tag %s for pull-mirror Repo[%d:%s/%s]: %w", tag.Name, repo.ID, repo.OwnerName, repo.Name, err)
	}
}
```

File: services/migrations/migrate.go [hunk 6]
```diff
@@ -482,6 +484,304 @@ func migrateRepository(doer *user_model.User, downloader base.Downloader, upload
...
+			if err = uploader.PatchReleases(releases[:relBatchSize]...); err != nil {
+				return err
+			}
+			releases = releases[relBatchSize:]
+		}
+
+		// Once all releases (if any) are inserted, sync any remaining non-release tags
+		// if err = uploader.SyncTags(); err != nil {
+		// 	return err
+		// }
```
Comment (harryzcy) [inline services/migrations/migrate.go]: Same as above

**Review by harryzcy — COMMENTED** [review]
Some existing code is removing usable release information entirely. This partially figures out the release syncing problem.

The change is introduced in #19125, which should be used when we only need tags, but not releases. However, it's clearly used wrongly.

**Comment by harryzcy** [comment]
I'm still trying to figure out migration related code. Looks like some code doing duplicated code, and some refactoring to migration code in general (not only this feature) might be needed. Maybe job for another PR after this one.

**Comment by lunny** [comment]
It's ready to review now?

**Comment by harryzcy** [comment]
@lunny almost.. almost.. let me go over everything again to see if anything is missed. I'll give an update later this week. I do want to finish it before 1.21 feature freezes.

File: models/issues/label.go (hunk 4 above)
Comment (delvh) [reply models/issues/label.go]: > There are lots of code using the default context, so I'm inclined to keep it.

Not for new functions. Only for existing functions.

File: models/issues/label.go [hunk 7]
```diff
@@ -232,19 +238,27 @@ func UpdateLabel(l *Label) error {
...
 		return err
 	}
+	defer committer.Close()
 
-	ctx, committer, err := db.TxContext(db.DefaultContext)
+	if err = deleteLabel(ctx, id, labelID); err != nil {
+		return err
+	}
+
+	return committer.Commit()
```
Comment (delvh) [inline models/issues/label.go]: `db.WithTx`

File: web_src/js/features/repo-migration.js [hunk 8]
```diff
@@ -45,7 +45,8 @@ function checkItems(tokenAuth) {
     enableItems = $user.val() !== '' || $pass.val() !== '';
   }
   if (enableItems && $service.val() > 1) {
-    if ($mirror.is(':checked')) {
+    const allowedServices = [2]; // services that supports migration items with mirroring
```
Comment (delvh) [inline web_src/js/features/repo-migration.js]: 1. What is `2`? `Gitea`? If yes, please add a comment explaining this magic literal
2. I don't think here is the correct place to initialize this variable. Please extract it into a global scope variable.

File: modules/migration/uploader.go [hunk 9]
```diff
@@ -17,6 +17,14 @@ type Uploader interface {
 	CreateComments(comments ...*Comment) error
 	CreatePullRequests(prs ...*PullRequest) error
 	CreateReviews(reviews ...*Review) error
+	UpdateTopics(topic ...string) error              // update topics of a repository, and delete unused ones
+	UpdateMilestones(milestones ...*Milestone) error // update milestones of a repository, and delete unused ones
+	UpdateLabels(labels ...*Label) error             // rewrite all issue labels and delete unused ones
```
Comment (delvh) [inline modules/migration/uploader.go]: I am unsure what to expect from these methods: Do they delete every label/milestone/… that is not present in the arguments?
Do the parameters contain if the corresponding label/milestone/… should now be deleted? (Unlikely)

Either way, I recommend a different approach: Simply let it be a no-op for things that have not changed.
Only update things that actually changed.
That has the benefit that users can still use mirror repos to create issues/labels/…, everything that does not change the code.
It will complicate the logic as you need to find out which of these things already exist and which were updated, but apart from that it makes a lot more sense semantically.

I don't see why we shouldn't do the `Patch` approach you describe below.

File: modules/migration/review.go [hunk 10]
```diff
@@ -42,6 +42,7 @@ type Review struct {
 	CreatedAt    time.Time `yaml:"created_at"`
 	State        string    // PENDING, APPROVED, REQUEST_CHANGES, or COMMENT
 	Comments     []*ReviewComment
+	OriginalID   int64
```
Comment (delvh) [inline modules/migration/review.go]: No comment here?

File: modules/migration/review.go [hunk 11]
```diff
@@ -52,16 +53,17 @@ func (r *Review) GetExternalID() int64 { return r.ReviewerID }
...
+	TreePath   string `yaml:"tree_path"`
+	DiffHunk   string `yaml:"diff_hunk"`
+	Position   int
+	Line       int
+	CommitID   string `yaml:"commit_id"`
+	PosterID   int64  `yaml:"poster_id"`
+	Reactions  []*Reaction
+	CreatedAt  time.Time `yaml:"created_at"`
+	UpdatedAt  time.Time `yaml:"updated_at"`
+	OriginalID int64
```
Comment (delvh) [inline modules/migration/review.go]: Or here?

File: services/mirror/mirror_pull.go [hunk 12]
```diff
@@ -332,7 +335,7 @@ func runSync(ctx context.Context, m *repo_model.Mirror) ([]*mirrorSyncResult, bo
 		log.Error("SyncMirrors [repo: %-v]: failed to update size for mirror repository: %v", m.Repo, err)
 	}
 
-	if m.Repo.HasWiki() {
+	if m.Repo.HasWiki() && m.SyncWiki {
```
Comment (delvh) [inline services/mirror/mirror_pull.go]: Won't this be pretty breaking for all existing mirrors?

File: models/repo/mirror.go [hunk 13]
```diff
@@ -25,6 +25,14 @@ type Mirror struct {
 	Interval    time.Duration
 	EnablePrune bool `xorm:"NOT NULL DEFAULT true"`
 
+	SyncWiki         bool
+	SyncIssues       bool
+	SyncMilestones   bool
+	SyncLabels       bool
+	SyncReleases     bool
+	SyncComments     bool
+	SyncPullRequests bool
```
Comment (delvh) [inline models/repo/mirror.go]: By the way, where is the migration for all these new options and the `OriginalID` fields?

Comment (delvh) [reply models/repo/mirror.go]: Furthermore, should we also reserve a boolean for `SyncProjects` already?
Because I guess that will also be a possible migration item in the future (although the API is not good enough yet to implement it already)

File: models/repo/repo.go [hunk 14]
```diff
@@ -558,12 +558,18 @@ func (repo *Repository) CanCreateBranch() bool {
...
 // AllowsPulls returns true if repository meets the requirements of accepting pulls and has them enabled.
 func (repo *Repository) AllowsPulls() bool {
-	return repo.CanEnablePulls() && repo.UnitEnabled(db.DefaultContext, unit.TypePullRequests)
+	return repo.CanEnablePulls() && !repo.IsMirror && repo.UnitEnabled(db.DefaultContext, unit.TypePullRequests)
+}
+
+// AllowsIssues returns true if repository meets the requirements of accepting issues and has them enabled.
+func (repo *Repository) AllowsIssues() bool {
+	// TODO: disable only when issues are synced
+	return !repo.IsMirror && repo.UnitEnabled(db.DefaultContext, unit.TypePullRequests)
```
Comment (delvh) [inline models/repo/repo.go]: I disagree with disallowing issues for mirror repos.
Why should they not be possible there?

File: services/mirror/mirror_pull.go [hunk 15]
```diff
@@ -411,6 +414,54 @@ func runSync(ctx context.Context, m *repo_model.Mirror) ([]*mirrorSyncResult, bo
...
+		PullRequests:   m.SyncPullRequests,
+		// Topics
+		// ReleaseAssets
+		MigrateToRepoID: repo.ID,
+		MirrorInterval:  m.Interval.String(),
+	}
+
+	_, err := migrations.SyncRepository(ctx, m.Repo.MustOwner(ctx), repo, opts, nil, lastSynced)
+	if err != nil {
+		log.Error("SyncMirrors [repo: %-v]: failed to sync repository: %v", m.Repo, err)
```
Comment (delvh) [inline services/mirror/mirror_pull.go]: SyncMirrors?

File: services/mirror/mirror_pull.go [hunk 16]
```diff
@@ -411,6 +414,54 @@ func runSync(ctx context.Context, m *repo_model.Mirror) ([]*mirrorSyncResult, bo
...
+		return false
+	}
+
+	return true
```
Comment (delvh) [inline services/mirror/mirror_pull.go]:
```suggestion
	}
	return err == nil
```

File: services/mirror/mirror_pull.go [hunk 17]
```diff
@@ -430,18 +481,28 @@ func SyncPullMirror(ctx context.Context, repoID int64) bool {
...
 	if !ok {
 		if err = repo_model.TouchMirror(ctx, m); err != nil {
 			log.Error("SyncMirrors [repo: %-v]: failed to TouchMirror: %v", m.Repo, err)
 		}
 		return false
 	}
 
+	if ok := runSyncMisc(ctx, m, lastSynced); !ok {
+		if err = repo_model.TouchMirror(ctx, m); err != nil {
+			log.Error("SyncMirrors [repo: %-v]: failed to TouchMirror: %v", m.Repo, err)
```
Comment (delvh) [inline services/mirror/mirror_pull.go]: Again, `SyncMirrors`?

File: models/repo/topic.go [hunk 18]
```diff
@@ -260,6 +260,34 @@ func AddTopic(repoID int64, topicName string) (*Topic, error) {
...
+		}
+
+		_, err := addTopicByNameToRepo(ctx, repoID, topicName)
+		if err != nil {
+			return err
+		}
+	}
+
+	if _, err := sess.ID(repoID).Cols("topics").Update(&Repository{
+		Topics: topicNames,
```
Comment (delvh) [inline models/repo/topic.go:278]: So you delete all existing topics that are not contained in the newly synced names?
That's neither what the function name indicates, nor what I would expect.

File: web_src/js/features/repo-migration.js
Comment (delvh) [inline web_src/js/features/repo-migration.js:1]: I am quite confused by this PR:
You do not offer any way to set if issues/pulls/milestones/… should be synced.
Is that planned as admins should enter that manually into the database for now, or how is that intended?

**Review by delvh — DISMISSED** [review]
I've looked through the easy 30%-40% now, still I found quite some problems.

File: services/mirror/mirror_pull.go (hunk 15 above)
Comment (harryzcy) [reply services/mirror/mirror_pull.go]: Do you mean it should be the correct function name here? like `runSyncMisc`?

File: models/repo/topic.go (hunk 18 above)
Comment (harryzcy) [reply models/repo/topic.go:278]: That's what this PR does right now.
What would you expect it to do? I'm happy to change that.
I would think that topics on Gitea would usually be the same as topics on a remote source like GitHub.

File: web_src/js/features/repo-migration.js
Comment (harryzcy) [reply web_src/js/features/repo-migration.js:1]: @delvh It's set when initializing the migration through the UI. I just added a screenshot to the description of this PR.
The options already exists. It's just previously all migration items are not available when it's a mirror, but now they are available to be toggled.

File: models/repo/mirror.go (hunk 13 above)
Comment (harryzcy) [reply models/repo/mirror.go]: Right now, the setting can only be set when starting a migration. Thus, existing mirrors have no way of turning on the feature in this PR.
I'm going to set those columns to default `false`, so no migration is necessary.

File: models/repo/repo.go (hunk 14 above)
Comment (harryzcy) [reply models/repo/repo.go]: This PR only implements on-way sync, that is from GitHub to Gitea. Allowing issues on Gitea itself will cause consistency issues. i.e. if there are Issue index 25 on both Github and Gitea, would will you do? There's no synchronization mechanism available across platforms.

File: modules/migration/uploader.go (hunk 9 above)
Comment (harryzcy) [reply modules/migration/uploader.go]: > Do they delete every label/milestone/… that is not present in the arguments?

Yes, that's the behavior right now.

Because there's no way to know what's deleted upstream. For Issues and PRs, they are add only, so it makes sense to keep everything. But topics, milestones, labels are different. So to delete things in Gitea, I can only delete anything not present in the arguments. And I think that's the behavior most people want. e.g. if you delete a milestone on GitHub, you would expect it to be deleted on Gitea as well, and don't want to go to Gitea again to delete the same thing.

In the future, we probably can support two-way syncing, but that's not the scope of this PR.

**Comment by mpeter50** [comment]
You don't have to push commits one by one. I think its fine e.g. at the end of the day, or when you are waiting for review.

Also, sometimes it might be useful to squash multiple commits into a single one. This happens with the rebase function of git, but if you are not familiar with it, you dont have to do it.

**Comment by harryzcy** [comment]
@lunny It's now ready for review

**Comment by lunny** [comment]
> @lunny It's now ready for review

Hi, lint error need to be fixed.

File: cmd/admin.go [hunk 19]
```diff
@@ -134,7 +134,8 @@ func runRepoSyncReleases(_ *cli.Context) error {
 			}
 			log.Trace(" currentNumReleases is %d, running SyncReleasesWithTags", oldnum)
 
-			if err = repo_module.SyncReleasesWithTags(ctx, repo, gitRepo); err != nil {
+			tagOnlyReleases := false
+			if err = repo_module.SyncReleasesWithTags(ctx, repo, gitRepo, tagOnlyReleases); err != nil {
```
Comment (lunny) [inline cmd/admin.go]: Why we need a local variable here?

Comment (harryzcy) [reply cmd/admin.go]: I can remove it. Previously, I thought it might be unclear if I just put a boolean there.

**Comment by DariuszGarbarz** [comment]
So what is the status? Is there any chance that it will be ready for 1.22?

**Comment by harryzcy** [comment]
> So what is the status? Is there any chance that it will be ready for 1.22?

Waiting for reviews

**Comment by cloning5480** [comment]
@delvh Would you be able to take a look at this PR again?

File: models/issues/issue.go [hunk 20]
```diff
@@ -922,3 +922,86 @@ func insertIssue(ctx context.Context, issue *Issue) error {
...
+		}
+		return nil
+	})
+}
+
+func upsertIssue(ctx context.Context, issue *Issue) (isInsert bool, err error) {
+	sess := db.GetEngine(ctx)
+
+	var issueIDs []int64
+	err = sess.Table("issue").Where("repo_id = ? AND `index` = ?", issue.RepoID, issue.Index).Cols("id").Find(&issueIDs)
```
Comment (lunny) [inline models/issues/issue.go]: You can use `Get` directly.

File: models/repo/mirror.go (hunk 13 above)
Comment (delvh) [reply models/repo/mirror.go]: The problem is, we have to maintain consistency.
If someone changes one of these columns in the future, the migration will fail for people who did not install any version with the original column. For example, assume we decide in 1.27 to change `SyncIssues` from `bool DEFAULT FALSE` to `int DEFAULT 0`, in which case only people who already installed Gitea in 1.22 can run this migration.
That's why all database changes need a migration.

File: models/repo/repo.go (hunk 14 above)
Comment (delvh) [reply models/repo/repo.go]: Okay, I can understand the synchronization problem.
However, three things:
1. it is a breaking change for anyone using a mirror to report issues on already.
2. The method name can also be read as `don't show issues tab` which is not its intended function, better: `AllowsCreatingIssues`
3. `unit.TypePullRequests`? Not `unit.TypeIssues`?

**Comment by rcarmo** [comment]
Does this include mirroring release packages as well? Right now Gitea only mirrors the tarballs, and any artifacts (like .appimages) added to GitHub releases are lost.

**Comment by harryzcy** [comment]
> Does this include mirroring release packages as well? Right now Gitea only mirrors the tarballs, and any artifacts (like .appimages) added to GitHub releases are lost.

Yes, included in this PR

**Comment by siddarthkay** [comment]
This PR looks cool, I'll try a build from this branch locally to see if this fixes our needs for having a mirror set up for Github repositories that also syncs, issues, comments on issues and PRs.
Thanks @harryzcy 
Looking forward to this PR getting vetted and merged in the future.

**Comment by harryzcy** [comment]
@siddarthkay Thank you for helping testing this out.

**Comment by siddarthkay** [comment]
hmm @harryzcy : I used your branch url as the source of my docker image and now I see that the options to select Issues, PRs while setting up a mirror are disabled.

Is it intended to be that way?

image

Here is my `docker-compose.yml` Incase it helps to replicate the issue.

```yml

version: "3"

networks:
  gitea:
    external: false

services:
  server:
    build:
      context: https://github.com/harryzcy/gitea.git#sync-issue-pr-and-more
      dockerfile: Dockerfile
    container_name: gitea
    environment:
      - USER_UID=1000
      - USER_GID=1000
      - GITEA__database__DB_TYPE=mysql
      - GITEA__database__HOST=db:3306
      - GITEA__database__NAME=gitea
      - GITEA__database__USER=gitea
      - GITEA__database__PASSWD=gitea
    restart: always
    networks:
      - gitea
    volumes:
      - ./gitea:/data
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
    ports:
      - "3000:3000"
      - "222:22"
    depends_on:
      - db

  db:
    image: mysql:8
    restart: always
    environment:
      - MYSQL_ROOT_PASSWORD=gitea
      - MYSQL_USER=gitea
      - MYSQL_PASSWORD=gitea
      - MYSQL_DATABASE=gitea
    networks:
      - gitea
    volumes:
      - ./mysql:/var/lib/mysql

```

**Comment by ajvpot** [comment]
Thanks so much for your work on this PR! I've set it up to attempt to see if it'll work for us.

The initial import worked fine but I'm encountering this error on subsequent syncs:

```
gitea-1        | 2024/05/24 23:02:23 ...eb/routing/logger.go:102:func1() [I] router: completed GET /ORGNAME/REPONAME/settings for ...:0, 200 OK in 15.1ms @ setting/setting.go:95(setting.Settings)
db-1           | 2024-05-24 21:02:24.317 UTC [35] ERROR:  duplicate key value violates unique constraint "issue_index_pkey"
db-1           | 2024-05-24 21:02:24.317 UTC [35] DETAIL:  Key (group_id)=(1452) already exists.
db-1           | 2024-05-24 21:02:24.317 UTC [35] STATEMENT:  INSERT INTO issue_index (group_id, max_index) VALUES ($1, $2)
db-1           | 2024-05-24 21:02:24.317 UTC [35] ERROR:  current transaction is aborted, commands ignored until end of transaction block
db-1           | 2024-05-24 21:02:24.317 UTC [35] STATEMENT:  SELECT max_index FROM issue_index WHERE group_id=$1
gitea-1        | 2024/05/24 23:02:24 ...irror/mirror_pull.go:465:runSyncMisc() [E] runSyncMisc [repo: <Repository 1452:ORGNAME/REPONAME>]: failed to run SyncRepository: pq: current transaction is aborted, commands ignored until end of transaction block
```

**Comment by rremer** [comment]
Hey @harryzcy are you actively working on this PR to sync pull request data from github->gitea, or are you looking for someone to pick up the torch? My company is interested in this feature, and I could:
a) develop off this branch and keep your commits, then open a new pr that has your commits + whatever fixes I needed
b) advocate we merge some portion of this, like schema changes and client/api/types without implementations and then I develop off of that
c) abandon this branch and close this (now draft) PR

This was bumped from 1.23 release goals to 1.24, I'd like to pitch in to make that happen but also don't want to step on your toes if you've got this.

**Comment by menacing-gingerbread** [comment]
Without wanting to sound impatient, given the time that has passed since this PR was last active, and that @harryzcy appears to have been inactive on this project for a significant period, it would be beneficial to clarify the next steps. To help maintain momentum on this feature.

**Comment by gmelodie** [comment]
I'd love to know what's the plan for next steps regarding this

**Comment by poloskey** [comment]
Perhaps it's better to close this issue since no action has been taken for some time and all promised milestones have not been met.

**Comment by GlassedSilver** [comment]
> Perhaps it's better to close this issue since no action has been taken for some time and all promised milestones have not been met.

And what would that achieve?

**Comment by AmirHmZz** [comment]
@harryzcy @lunny Any updates on this PR? This feature is really useful in some of Gitea use cases.

**Comment by GlassedSilver** [comment]
> @harryzcy @lunny Any updates on this PR? This feature is really useful in some of Gitea use cases.

essential even

If you wanna make backups of repos you rely on juuust in case something happens to them you wanna have a known-good copy of code, issues (which often contain guidance for setup, etc) and maybe still WIP enhancements for PRs.

As it is now this stuff typically dies unless someone invested a lot of manual work.

**Comment by harryzcy** [comment]
> @harryzcy @lunny Any updates on this PR? This feature is really useful in some of Gitea use cases.

Sorry that I'm not active on this PR and this repo right now, limited by my bandwidth. I may be able to pick this up sometime but I cannot provide much guarantee for now.

If anyone want to take over this PR, or a subset of this PR (including cherry-picking off the code I committed), you are welcome to. Maybe this PR is too large of a change that's worth breaking up into multiple PRs as well.

**Comment by davidlovas** [comment]
@harryzcy, picking this up per your offer to break it up, and thanks for reviewing the initial PRs I opened. I've since updated the PRs, with the full plan shared on #18369. Continuing there to keep the discussion in one place.

**Comment by bircni** [comment]
This PR had no activity for over a year, has lots of conflicts now, has unanswered feedback or is otherwise abandoned and will thus be closed.
If you want to continue on this PR, reopen it, fix all merge conflicts, and answer the existing feedback, if it exists.

**Comment by davidlovas** [comment]
> This PR had no activity for over a year, has lots of conflicts now, has unanswered feedback or is otherwise abandoned and will thus be closed.
> 
> If you want to continue on this PR, reopen it, fix all merge conflicts, and answer the existing feedback, if it exists.

I already did all that, read my comments on the issue please.  It starts with the graphql PR which is open, and there are another 3 PRs which are stacked on top of that one.

**Comment by AmirHmZz** [comment]
@bircni @delvh any updates on reviewing this?

**Comment by davidlovas** [comment]
To clarify where this stands for everyone following along:

I have re-derived this feature against current main as a stack of smaller PRs. The first is #38694, a GraphQL downloader for GitHub migrations, and it has to land first because without it this feature isn't realistic: over REST every comment, review and reaction is its own request, so mirroring a repo this size means hundreds of thousands of calls and exhausts rate limits no matter how many tokens you use. GraphQL brings an issue with its comments and reactions back in a single request. The mirror sync itself is built and working on top of it, and I'll open those PRs once #38694 lands instead of putting them through months of rebases first.

Meanwhile:
- the follow-up PRs can be previewed on my fork, stacked on the #38694 branch: [davidlovas/gitea#2](https://github.com/davidlovas/gitea/pull/2), [#3](https://github.com/davidlovas/gitea/pull/3), [#4](https://github.com/davidlovas/gitea/pull/4)
- or build and run the whole thing today from the [`feature/metadata-mirror`](https://github.com/davidlovas/gitea/tree/feature/metadata-mirror) branch (rebased onto main regularly, expect force pushes)
- a live instance mirrors go-gitea itself with issues, PRs, comments, reviews, labels and reactions kept current: https://gitea.rexovas.com/rexovas/gitea

The stack is maintained in a private fork with automation that keeps it rebased against upstream and tested continuously; these public branches are snapshots of that and may lag at any given moment.

Given that, I think this PR can be closed in favor of that sequence.

Related issues:
- #18369 (feature proposal)
- #13243 (migration rate limits)
